### DISCUSSION TITLE
Surprising difference between `f_a(x)` and `f_1(x)`

### DISCUSSION THREAD

**Discussion by NicoWeio** [typst/typst#2086] [General]
### Description

In my field, both letters and numbers regularly appear as subscript (/superscript).
Coming from LaTeX, I was surprised that Typst handles `f_a(x)` and `f_1(x)` differently:

#### Typst:

```
- $f_2(x)$
- $f_g(x)$
- $f^2(x)$
- $f^g(x)$
```
![typst-ex-2](https://github.com/typst/typst/assets/12870528/ef35152f-516b-4f23-b342-2ec0c894f514)

#### KaTeX (but LaTeX would do the same):
```
f_2(x) \\
f_g(x) \\
f^2(x) \\
f^g(x)
```
![typst-ex1](https://github.com/typst/typst/assets/12870528/17db0026-0dce-4ef2-a008-a8edf9590769)

Similar discussions have been held regarding primes (see #1908 / #1614, for example), but I think this deserves a separate issue: Is this something one has to live with, or could we choose to modify the current behavior? We could consider what's frequently used:
- $f_2(x)$, $f_g(x)$ are common
- $f_{2(x)}$ usually doesn't make sense
- $f_{g(x)}$ is AFAICT **much less common** than $f_g(x)$

### Reproduction URL

https://typst.app/project/r8tWok4Rf9DFrtLXe_hHMy

### Operating system

Web app, Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enivex** [comment]
This was an intentional change to make fewer parentheses needed (`e^f(x)` vs `e^(f(x))`), at the cost of sometimes needing to add a space to disambiguate. That is `a_n (x)`, for instance.

Neither solution is ideal.

See https://github.com/typst/typst/pull/985

**Comment by laurmaedje** [comment]
This comes up very frequently. The latest idea was to make an overview of all cases, then look how frequent things are are. If we have a clear idea what notation is affected by each policy and how common it is, we can make an informed decision on the final approach.

**Reply by francois-rozet** [reply]
In my fields (machine learning, optimization, calculus, ...) $f_g(x)$ is MUCH more frequent than $f_{g(x)}$.

**Reply by JuliDi** [reply]
I find it very annoying that I can write `x(t)` and `x_1(t)`, but `x_a(t)` will suddenly render differently. Adding an extra space in _some_ cases looks pretty inconsistent in the code, too.

My main argument for not putting the term in brackets as a sub-/superscript is that (to my knowledge) also no programming language considers brackets as part of the identifier. You can always do `<ident>(x)` and `(x)` would never be part of the function name depending on whether you have `foo_a(x)` or `foo_1(x)`. Also, it would feel like it's more consistent with Latex.

**Reply by laurmaedje** [reply]
It is planned to change it.

**Reply by hjalmarlucius** [reply]
Has there been any progress on this? I love typst but this simplification seems to go beyond what is possible to simplify and inconsistency ensues.

**Comment by MHellmund** [comment]
In my experience, functions with indices $f_i(x)$ or vector fields like $\psi_\alpha(x,t),  A^\mu(x,t)$  are much more frequent than $e^{f(x)}$. 

When I need a complex expression in the exponent, I may write it for better readability as $\exp(f(x,y,z))$, anyway.  

Spaces between a function and its argument list `f_i (x,t)` look ugly  in my eyes.    

Also, from a mathematical POV, having different  rules for a numeric index $f_2(x)$ and a symbolic index $f_i(x)$ feels a bit odd.

**Comment by yaksher** [comment]
What about having some kind of "attach" constant that is basically infix parenthesis? So `e^f attach (x)` is `e^(f(x))` and `f_i(x)` is `f_i (x)`?

It would need to have a short name of some kind, perhaps a single non-letter character. ~ is relatively convenient to type and the vast majority of the time when it's used in math, you want it to be a relation anyway, for which you use `tilde`. So `e^f~(x)` or whatever.

**Comment by MHellmund** [comment]
If this is the place for half-baked ideas: 
Allow multi letter sequences immediately after sub/superscript. 
So we can write $A_{ij}$ as `A_ij` instead of `A_(i j)`, avoiding parentheses and empty spaces.

**Reply by NicoWeio** [reply]
I agree that it would be convenient, but IMO the cost in terms of consistency is too high. It “opens the door” to eventually allowing multi-letter sequences everywhere, which I'm not a fan of.
Plus, if you repeatedly use the same indices, you can define an alias like `#let ij = $i j$`.

**Comment by michihupf** [comment]
I am strongly against this `f_g (x)` with a space, clearly means what you are looking for whereelese `f_g(x)` is open to interpretation as this could mean your case or g is a function that gives an index (i know this is quite infrequent).
In my opinion what makes typst so great is that you can clearly read the source code and understand everything, where latex is sometimes a bit convoluted. I don't think it would be in the spirit of this project to just reimplement latex but to reimagine it. Considering `f_g(x)` is less messy than `f_(g(x))` all for the cost of one space in the other case I'd say the current version is perfectly fine.

**Comment by NicoWeio** [comment]
ICYMI, there have been some recent developments:
#6442 would have resolved the issue in question but was reverted in #7084.
